### PR TITLE
Supermicro X11 BIOS, BMC firmware install support

### PR DESCRIPTION
**Pull request by joelrebel**
## What does this PR implement/change/remove?

Implements firmware install support for Supermicro X11s through its XML API.

### Checklist
- [X] Tests added
- [X] Similar commits squashed

### The HW vendor this change applies to (if applicable)
Supermicro

### The HW model number, product name this change applies to (if applicable)

It's likely this works on all X11s, although, for now this code has been tested on and limits itself to these models.

- X11DPH-T
- X11SCM-F

## Description for changelog/release notes

```
Implement firmware install for Supermicro BIOS and BMC.
```

### CODE REVIEW

File: examples/install-firmware/doc.go [hunk 1]
```diff
@@ -2,23 +2,57 @@
...
-		-user string
-					Username to login with
-		-version string
-					The firmware version being installed
+Note: The example installs the firmware and polls until the status until the install is complete,
+and if required by the install process - power cycles the host.
+
+							$ go run ./examples/v1/install-firmware/main.go -h
+							Usage of /tmp/go-build2950657412/b001/exe/main:
+								-cert-pool string
```
Comment (splaspood) [inline examples/install-firmware/doc.go:10]: It looks like all these flags have a single dash, but should have two?

File: examples/install-firmware/main.go [hunk 2]
```diff
@@ -20,24 +22,33 @@ func main() {
 	user := flag.String("user", "", "Username to login with")
 	pass := flag.String("password", "", "Username to login with")
 	host := flag.String("host", "", "BMC hostname to connect to")
+	component := flag.String("component", "", "Component to be updated (bmc, bios.. etc)")
 	withSecureTLS := flag.Bool("secure-tls", false, "Enable secure TLS")
 	certPoolPath := flag.String("cert-pool", "", "Path to an file containing x509 CAs. An empty string uses the system CAs. Only takes effect when --secure-tls=true")
 	firmwarePath := flag.String("firmware", "", "The local path of the firmware to install")
-	firmwareVersion := flag.String("version", "", "The firmware version being installed")
+	//firmwareVersion := flag.String("version", "", "The firmware version being installed")
```
Comment (splaspood) [inline examples/install-firmware/main.go:29]: Why is this commented out?

File: examples/install-firmware/main.go [hunk 3]
```diff
@@ -53,40 +64,81 @@ func main() {
...
+	for {
+		if ctx.Err() != nil {
+			l.Fatal(ctx.Err())
+		}
+
+		state, err := cl.FirmwareInstallStatus(ctx, "", *component, "")
+		if err != nil {
+			// when its under update a connection refused is returned
+			if strings.Contains(err.Error(), "connection refused") || strings.Contains(err.Error(), "operation timed out") {
+				l.Info("BMC refused connection, BMC most likely resetting..")
```
Comment (splaspood) [inline examples/install-firmware/main.go:96]: small nit, add a third period: "resetting..."

**Review by splaspood — CHANGES_REQUESTED** [review]

**Comment by joelrebel** [comment]
Thanks for the feedback @splaspood, I had made some changes to the example to run specifically on SMC hardware. Its now fixed up to be more generic.

**Review by splaspood — APPROVED** [review]

File: examples/install-firmware/doc.go (hunk 1 above)
Comment (joelrebel) [reply examples/install-firmware/doc.go:10]: The flags package accepts both single and double dashes, I updated the examples to be consistent.

**Pull request merged**